sam/pileup.go

```diff
@@ -0,0 +1,366 @@
+			send <- pb.buf[i]
+		}
+		pb.incrementIdx()
+		resetPile(&pb.buf[i])
+		return
+	}
+	if pb.buf[i].Pos >= s.Pos {
+		pb.idx = i
+		return true
+	}
+	if (pb.buf[i].touched && passesPileFilters(pb.buf[i], pileFilters)) || includeNoData {
+		send <- pb.buf[i]
+	}
+	pb.incrementIdx()
+	resetPile(&pb.buf[i])
+	return
+}
+
+// addBase to the pileBuffer at the input position
+func (pb *pileBuffer) addBase(refidx int, pos uint32, base dna.Base) {
+	pile := pb.getIdx(refidx, pos)
+	pile.Count[base]++
+	pile.touched = true
+}
+
+// addIns to the pileBuffer at the input position
+func (pb *pileBuffer) addIns(refidx int, pos uint32, seq string) {
+	pile := pb.getIdx(refidx, pos)
+	pile.InsCount[seq]++
+	pile.touched = true
+}
+
+// getIdx returns a pointer to the pile at index pos in buf
+func (pb *pileBuffer) getIdx(refidx int, pos uint32) *Pile {
+	if pos < pb.buf[pb.idx].Pos {
+		log.Panic("tried to retrieve past position. unsorted input?")
+	}
+	queryIdx := int(pos-pb.buf[pb.idx].Pos) + pb.idx
+
+	// calc if queryIdx wraps around to start of buffer
+	if queryIdx >= len(pb.buf) {
+		queryIdx -= len(pb.buf)
+	}
+
+	// check position for correct data, return if found
+	if queryIdx < len(pb.buf) && pb.buf[queryIdx].Pos == pos {
+		return &pb.buf[queryIdx]
+	}
+
+	// at this point, one of the following is true:
+	// 1: the buffer is empty and needs to be filled
+	// 2: the buffer is partially full and can be filled to pos
+	// 3: the buffer is full and needs to be expanded
+
+	switch {
+	case pb.buf[pb.idx].RefIdx == -1: // buffer is empty (case 1)
+		pb.initializeFromEmpty(pos, refidx)
+		return &pb.buf[pb.idx]
+
+	case pb.buf[pb.decrementIdx()].RefIdx == -1: // buffer is partially full (case 2)
+		pb.fillFromPartial()
+		return pb.getIdx(refidx, pos)
+
+	default: // buffer is full (case 3)
+		pb.expand()
+		return pb.getIdx(refidx, pos)
+	}
+
+}
+
+// initializeFromEmpty fills an empty pileBuffer starting with the position of buf.Seq[0]
+func (pb *pileBuffer) initializeFromEmpty(pos uint32, refidx int) {
+	for ; pb.buf[pb.idx].RefIdx == -1; pb.idx = pb.incrementIdx() {
+		pb.buf[pb.idx].Pos = pos
+		pb.buf[pb.idx].RefIdx = refidx
+		pos++
+	}
+}
+
+// fillFromPartial fills a partially filled pileBuffer
+func (pb *pileBuffer) fillFromPartial() {
+	start := pb.idx
+	refidx := pb.buf[pb.idx].RefIdx
+	pos := pb.buf[pb.idx].Pos + 1
+	for pb.idx = pb.incrementIdx(); pb.idx != start; pb.idx = pb.incrementIdx() {
+		if pb.buf[pb.idx].RefIdx == -1 {
+			pb.buf[pb.idx].RefIdx = refidx
+			pb.buf[pb.idx].Pos = pos
+		}
+		pos++
+	}
+}
+
+// expand the pileBuffer by 2x its current size
+func (pb *pileBuffer) expand() {
+	newpb := newPileBuffer(len(pb.buf) * 2)
+	for i := 0; i < len(pb.buf); i++ {
+		newpb.buf[i] = pb.buf[pb.idx]
+		pb.idx = pb.incrementIdx()
+	}
+	*pb = *newpb
+}
+
+// incrementIdx returns the idx of the pile after pb.idx
+func (pb *pileBuffer) incrementIdx() int {
+	newidx := pb.idx + 1
+	if newidx == len(pb.buf) {
```

Comment:
Now I am pretty sure this is a circular buffer.  I think there a decent amount of added complexity here to implement the circularity.  What do you think about a linked list of Piles?  This way you can quickly move things from the front to the end (I think this is the reason you made it circular), but you still end up with a linear data structure, so you don't need the code to deal with where the start is located (always at head) and you don't need the code to skip from the end to the beginning.